File: packages/adapters/src/langchain.ts
/**
 * SkillForge LangChain Adapter
 * 
 * Captures failures from LangChain chains, agents, and tools.
 */

import { SkillForge, Failure } from '@skillforge/sdk';

export class LangChainAdapter {
  private forge: SkillForge;

  constructor(forge: SkillForge) {
    this.forge = forge;
  }

  /**
   * Wrap a LangChain Runnable (chains, agents, etc.)
   */
  wrapRunnable(runnable: any): any {
    const originalInvoke = runnable.invoke?.bind(runnable);
    
    if (originalInvoke) {
      runnable.invoke = async (input: any, options?: any) => {
        try {
          return await originalInvoke(input, options);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            input,
            runnableType: this.getRunnableType(runnable),
            options,
          });
          throw error;
        }
      };
    }

    const originalBatch = runnable.batch?.bind(runnable);
    
    if (originalBatch) {
      runnable.batch = async (inputs: any[], options?: any) => {
        try {
          return await originalBatch(inputs, options);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            inputs,
            runnableType: this.getRunnableType(runnable),
            failedIndex: this.findFailedIndex(error, inputs.length),
          });
          throw error;
        }
      };
    }

    const originalStream = runnable.stream?.bind(runnable);
    
    if (originalStream) {
      runnable.stream = async function* (input: any, options?: any) {
        try {
          for await (const chunk of originalStream(input, options)) {
            yield chunk;
          }
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            input,
            runnableType: this.getRunnableType(runnable),
            streamError: true,
          });
          throw error;
        }
      }.bind(this);
    }

    return runnable;
  }

  /**
   * Wrap a LangChain Agent
   */
  wrapAgent(agent: any): any {
    // Wrap the agent executor
    if (agent.agentExecutor) {
      this.wrapRunnable(agent.agentExecutor);
    }

    // Wrap tools
    if (agent.tools) {
      this.wrapTools(agent.tools);
    }

    return agent;
  }

  /**
   * Wrap LangChain tools
   */
  wrapTools(tools: any[]): void {
    for (const tool of tools) {
      const originalCall = tool.call?.bind(tool);
      
      if (originalCall) {
        tool.call = async (input: any) => {
          try {
            return await originalCall(input);
          } catch (error: any) {
            await this.captureLangChainFailure(error, {
              toolName: tool.name,
              toolDescription: tool.description,
              input,
            });
            throw error;
          }
        };
      }

      // Some tools use _call (protected method)
      if (tool._call) {
        const originalUnderscoreCall = tool._call.bind(tool);
        tool._call = async (input: any) => {
          try {
            return await originalUnderscoreCall(input);
          } catch (error: any) {
            await this.captureLangChainFailure(error, {
              toolName: tool.name,
              input,
            });
            throw error;
          }
        };
      }
    }
  }

  /**
   * Wrap a LangChain Chain
   */
  wrapChain(chain: any): any {
    // Modern LangChain uses Runnable
    if (chain.invoke) {
      return this.wrapRunnable(chain);
    }

    // Legacy Chain interface
    const originalCall = chain.call?.bind(chain);
    
    if (originalCall) {
      chain.call = async (args: any) => {
        try {
          return await originalCall(args);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            chainType: chain.constructor.name,
            args,
          });
          throw error;
        }
      };
    }

    const originalRun = chain.run?.bind(chain);
    
    if (originalRun) {
      chain.run = async (input: string) => {
        try {
          return await originalRun(input);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            chainType: chain.constructor.name,
            input,
          });
          throw error;
        }
      };
    }

    return chain;
  }

  /**
   * Wrap a LangChain LLM
   */
  wrapLLM(llm: any): any {
    const originalGenerate = llm.generate?.bind(llm);
    
    if (originalGenerate) {
      llm.generate = async (prompts: string[], options?: any) => {
        try {
          return await originalGenerate(prompts, options);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            llmType: llm.constructor.name,
            promptCount: prompts.length,
            firstPrompt: prompts[0]?.slice(0, 100),
          });
          throw error;
        }
      };
    }

    const originalInvoke = llm.invoke?.bind(llm);
    
    if (originalInvoke) {
      llm.invoke = async (input: any, options?: any) => {
        try {
          return await originalInvoke(input, options);
        } catch (error: any) {
          await this.captureLangChainFailure(error, {
            llmType: llm.constructor.name,
            input: typeof input === 'string' ? input.slice(0, 100) : input,
          });
          throw error;
        }
      };
    }

    return llm;
  }

  /**
   * Get runnable type for debugging
   */
  private getRunnableType(runnable: any): string {
    if (runnable.lc_namespace) {
      return runnable.lc_namespace.join('.');
    }
    return runnable.constructor.name;
  }

  /**
   * Find which batch item failed
   */
  private findFailedIndex(error: any, total: number): number | null {
    // LangChain sometimes includes the index in error
    if (error.index !== undefined) {
      return error.index;
    }
    if (error.message?.match(/item (\d+)/)) {
      return parseInt(error.message.match(/item (\d+)/)[1]);
    }
    return null;
  }

  /**
   * Capture failure with LangChain-specific context
   */
  private async captureLangChainFailure(error: any, context: any): Promise<Failure> {
    return await this.forge.captureFailure(error, {
      task: context.input || context.args || 'LangChain execution',
      context: {
        framework: 'langchain',
        ...context,
        langchainVersion: this.getLangChainVersion(),
      },
    });
  }

  /**
   * Get LangChain version
   */
  private getLangChainVersion(): string {
    try {
      // Try to get from package
      const pkg = require('langchain/package.json');
      return pkg.version;
    } catch {
      return 'unknown';
    }
  }

  /**
   * Apply skill to LangChain chain/agent
   */
  async applySkill(skill: any, target: any): Promise<void> {
    // LangChain uses different mechanisms for different types
    if (target.tools) {
      // Agent - add as tool
      target.tools.push(this.skillToTool(skill));
    } else if (target.callbacks) {
      // Runnable - add callback
      // This is more complex, would need custom callback handler
    }
  }

  /**
   * Convert skill to LangChain tool
   */
  private skillToTool(skill: any): any {
    return {
      name: skill.name,
      description: skill.description,
      call: async (input: string) => {
        // Execute skill
        // In production, would call skill execution engine
        return `Applied skill ${skill.name}`;
      },
    };
  }

  /**
   * Get skill suggestions for LangChain context
   */
  async getSuggestions(errorType: string): Promise<any[]> {
    const langChainSkillMap: Record<string, string[]> = {
      'OutputParserException': ['structured-output-protocol', 'output-fixer'],
      'JSONDecodeError': ['json-sanitizer', 'robust-json-parser'],
      'ToolInputError': ['tool-call-fixer'],
      'TokenLimitError': ['token-optimizer', 'chunking-handler'],
      'RateLimitError': ['rate-limit-handler', 'retry-with-backoff'],
    };

    const skillNames = langChainSkillMap[errorType] || [];
    return skillNames.map(name => ({ name, framework: 'langchain' }));
  }
}

// Auto-detection helper
export function detectLangChain(obj: any): boolean {
  return (
    obj &&
    (obj.lc_namespace || obj.lc_id || 
     obj.constructor?.name?.includes('Chain') ||
     obj.constructor?.name?.includes('Agent'))
  );
}

export default LangChainAdapter;

File: packages/adapters/src/langchain.py
"""
SkillForge LangChain Adapter (Python version)

For Python-based LangChain applications.
"""

from typing import Any, Dict, List, Optional, Callable
from functools import wraps
import traceback

from skillforge import SkillForge, Failure


class LangChainAdapter:
    """
    Adapter for LangChain Python applications.
    
    Usage:
        from skillforge import SkillForge
        from skillforge.adapters import LangChainAdapter
        
        forge = SkillForge(agent='my-agent')
        adapter = LangChainAdapter(forge)
        
        # Wrap chain
        chain = adapter.wrap_chain(my_chain)
        
        # Wrap agent
        agent = adapter.wrap_agent(my_agent)
    """
    
    def __init__(self, forge: SkillForge):
        self.forge = forge
    
    def wrap_chain(self, chain: Any) -> Any:
        """Wrap a LangChain Chain"""
        # Try different invoke methods based on chain type
        
        if hasattr(chain, 'invoke'):
            # Modern Runnable interface
            original_invoke = chain.invoke
            
            @wraps(original_invoke)
            async def wrapped_invoke(input: Any, config: Optional[Dict] = None):
                try:
                    return await original_invoke(input, config)
                except Exception as error:
                    await self._capture_failure(error, {
                        'input': input,
                        'chain_type': type(chain).__name__,
                        'config': config,
                    })
                    raise
            
            chain.invoke = wrapped_invoke
        
        if hasattr(chain, 'acall'):
            # Legacy async interface
            original_acall = chain.acall
            
            @wraps(original_acall)
            async def wrapped_acall(inputs: Dict[str, Any]):
                try:
                    return await original_acall(inputs)
                except Exception as error:
                    await self._capture_failure(error, {
                        'inputs': inputs,
                        'chain_type': type(chain).__name__,
                    })
                    raise
            
            chain.acall = wrapped_acall
        
        if hasattr(chain, '__call__'):
            # Legacy sync interface
            original_call = chain.__call__
            
            @wraps(original_call)
            def wrapped_call(inputs: Dict[str, Any]):
                try:
                    return original_call(inputs)
                except Exception as error:
                    # Sync version - still capture but don't await
                    import asyncio
                    try:
                        loop = asyncio.get_event_loop()
                    except RuntimeError:
                        loop = asyncio.new_event_loop()
                        asyncio.set_event_loop(loop)
                    
                    loop.run_until_complete(self._capture_failure(error, {
                        'inputs': inputs,
                        'chain_type': type(chain).__name__,
                    }))
                    raise
            
            chain.__call__ = wrapped_call
        
        return chain
    
    def wrap_agent(self, agent: Any) -> Any:
        """Wrap a LangChain Agent"""
        # Wrap agent executor
        if hasattr(agent, 'agent_executor'):
            self.wrap_chain(agent.agent_executor)
        
        # Wrap tools
        if hasattr(agent, 'tools'):
            self.wrap_tools(agent.tools)
        
        return agent
    
    def wrap_tools(self, tools: List[Any]) -> List[Any]:
        """Wrap a list of LangChain tools"""
        for tool in tools:
            if hasattr(tool, '_run'):
                # StructuredTool
                original_run = tool._run
                
                @wraps(original_run)
                def wrapped_run(tool_input: str, original=original_run, tool_name=tool.name):
                    try:
                        return original(tool_input)
                    except Exception as error:
                        import asyncio
                        try:
                            loop = asyncio.get_event_loop()
                        except RuntimeError:
                            loop = asyncio.new_event_loop()
                            asyncio.set_event_loop(loop)
                        
                        loop.run_until_complete(self._capture_failure(error, {
                            'tool_name': tool_name,
                            'tool_input': tool_input,
                        }))
                        raise
                
                tool._run = wrapped_run
            
            if hasattr(tool, 'arun'):
                # Async tool
                original_arun = tool.arun
                
                @wraps(original_arun)
                async def wrapped_arun(tool_input: str, original=original_arun, tool_name=tool.name):
                    try:
                        return await original(tool_input)
                    except Exception as error:
                        await self._capture_failure(error, {
                            'tool_name': tool_name,
                            'tool_input': tool_input,
                        })
                        raise
                
                tool.arun = wrapped_arun
        
        return tools
    
    def wrap_llm(self, llm: Any) -> Any:
        """Wrap a LangChain LLM"""
        if hasattr(llm, 'agenerate'):
            original_agenerate = llm.agenerate
            
            @wraps(original_agenerate)
            async def wrapped_agenerate(prompts: List[str], **kwargs):
                try:
                    return await original_agenerate(prompts, **kwargs)
                except Exception as error:
                    await self._capture_failure(error, {
                        'llm_type': type(llm).__name__,
                        'prompt_count': len(prompts),
                        'first_prompt': prompts[0][:100] if prompts else None,
                    })
                    raise
            
            llm.agenerate = wrapped_agenerate
        
        return llm
    
    async def _capture_failure(self, error: Exception, context: Dict[str, Any]) -> Failure:
        """Capture failure with LangChain-specific context"""
        return await self.forge.capture_failure(
            error=error,
            context={
                'task': context.get('input') or context.get('inputs') or 'LangChain execution',
                'context': {
                    'framework': 'langchain',
                    'framework_version': self._get_langchain_version(),
                    'traceback': traceback.format_exc(),
                    **context,
                },
            }
        )
    
    def _get_langchain_version(self) -> str:
        """Get LangChain version"""
        try:
            import langchain
            return langchain.__version__
        except (ImportError, AttributeError):
            return 'unknown'
    
    async def get_suggestions(self, error_type: str) -> List[Dict[str, Any]]:
        """Get skill suggestions for LangChain context"""
        langchain_skill_map = {
            'OutputParserException': ['structured-output-protocol', 'output-fixer'],
            'JSONDecodeError': ['json-sanitizer', 'robust-json-parser'],
            'ToolInputError': ['tool-call-fixer'],
            'TokenLimitError': ['token-optimizer', 'chunking-handler'],
            'RateLimitError': ['rate-limit-handler', 'retry-with-backoff'],
            'ValidationException': ['input-validator', 'schema-enforcer'],
        }
        
        skill_names = langchain_skill_map.get(error_type, [])
        return [{'name': name, 'framework': 'langchain'} for name in skill_names]
    
    def skill_to_tool(self, skill: Any) -> Any:
        """Convert a skill to a LangChain tool"""
        from langchain.tools import BaseTool
        
        class SkillTool(BaseTool):
            name = skill.name
            description = skill.description
            
            def _run(self, tool_input: str) -> str:
                # Execute skill
                # In production, would call skill execution engine
                return f"Applied skill {skill.name}"
            
            async def _arun(self, tool_input: str) -> str:
                return self._run(tool_input)
        
        return SkillTool()


def detect_langchain(obj: Any) -> bool:
    """Detect if object is a LangChain component"""
    return (
        hasattr(obj, 'lc_namespace') or
        hasattr(obj, 'lc_id') or
        type(obj).__module__.startswith('langchain') or
        'Chain' in type(obj).__name__ or
        'Agent' in type(obj).__name__
    )
